Adds locale as a chainable operator to bar chart

maikdiepenbroek opened the pull request:
to enable passing a custom locale. This can be done in two forms: string (following the pattern: [a-z]{2}-[A-Z]{2} or an object in the form of a d3-format locale definition, f.e.: https://cdn.jsdelivr.net/npm/d3-format@1.4.2/locale/en-US.json. If a string is passed, it's locale is retrieved asynchronous and then passed to d3-format.formatDefaultLocale

DalerAsrorov commented on src/charts/bar.js:
```diff
@@ -164,7 +166,16 @@ define(function(require) {
-            _selection.each(function(_data) {
+            if (locale && enableLabels) {
+                setDefaultLocale(locale)
+                    .then(() => initChart(_selection))
```
Can initialization of the locale be done somewhere else?

maikdiepenbroek replied:
Sure, i first had this in the drawLabels call. What place suits you better ?

DalerAsrorov commented on src/charts/helpers/locale.js:
```diff
@@ -0,0 +1,29 @@
+define(function (require) {
+    const {FORMAT_LOCALE_URL} = require('./constants');
+    const {formatDefaultLocale} = require('d3-format');
```
There needs to be a test for this helper. Make sure that every use-case is covered.

maikdiepenbroek replied:
Just to provide some context, this PR in its current state is only to discuss the current approach. If we decide it is the way to go, i can add it to the other charts and update docs/tests

maikdiepenbroek commented:
@DalerAsrorov and @Golodhros As a side note, when i perform the linting script i get a lot of code-style changes, also i see a lot of unused variables coming from require etc. When this PR is in a later state, i'll create a separate commit that fixes linting issues etc.

Golodhros commented on src/charts/bar.js:
```diff
@@ -164,7 +166,16 @@ define(function(require) {
+        }
+
+        function initChart(selection) {
+            selection.each(function(_data) {
```
I would say to trigger the locale request here and keep a reference to the promise. Something like:
```
  localeRequest = fetchLocale();
  selection.each(....
```
Later on, whenever the locale is actually needed, we'll do:
```
  localeRequest.then(...
```

That way we'll trigger the request as soon as possible (if needed) while blocking the render of anything that depends on it. 

We should probably include a configurable timeout in case the request fails.

maikdiepenbroek replied:
Liked your suggestion, i've implemented it like this.

maikdiepenbroek commented:
@Golodhros and @DalerAsrorov  If you've got the time please review the current state of this PR. So far i've added jsdoc and unit tests for the locale helper. Also i added a small unit test for the locale property in the bart chart. Next to that i've implemented the suggestion of storing a reference to the getLocale promise. 
I'm having a difficult time getting to know the way of testing the complete bar chart, maybe you could help me with rounding of the tests especially mocking/spying. 

Once the bar chart is tested to standards,  i can move on and implement the locale property in the other charts (where applicable) and expand the docs.

Sounds like a plan ?

Golodhros commented on src/charts/bar.js:
```diff
@@ -9,13 +9,13 @@ define(function(require) {
     const d3Format = require('d3-format');
     const d3Scale = require('d3-scale');
     const d3Selection = require('d3-selection');
-    const d3Transition = require('d3-transition');
```
This needs to be loaded, even if we don't use it.

maikdiepenbroek replied:
Got it, i reversed it.

Golodhros replied:
Did you push the changes?

maikdiepenbroek replied:
No, did the changes locally only.

Golodhros commented on src/charts/bar.js:
```diff
@@ -1292,6 +1302,15 @@ define(function(require) {
             return this;
         };
 
```
We would need a comment on this accessor so we generate the documentation

maikdiepenbroek replied:
Added it

Golodhros commented on src/charts/bar.js:
```diff
@@ -177,7 +183,11 @@ define(function(require) {
 
-                if (enableLabels) {
+                if (enableLabels && locale) {
+                    localeRequest
```
Would it work if we set up localeRequest as a resolved promise by default?

maikdiepenbroek replied:
I'm not sure what you mean, could you explain with some (pseudo) code ?

Golodhros replied:
Instead of declaring localeRequest as undefined, you could do:
  localeRequest = Promise.resolve();

Then, you wouldn't need to add the branching here:

   localeRequest.then(() => drawLabels())
         .catch(() => new Error('You need an Internet connection in order to load the selected locale'))

maikdiepenbroek replied:
I've added the Promise.resolve to the declaration of localeRequest, but i don't quite understand the  branching part.

Golodhros replied:
If the value of localeRequest is a resolved promise by default, then you don't need to check for 'enableLables' or 'locale', you can just put the same code for all cases.

maikdiepenbroek replied:
> If the value of localeRequest is a resolved promise by default, then you don't need to check for 'enableLables' or 'locale', you can just put the same code for all cases.

Got it now, good catch!

Golodhros commented on src/charts/helpers/locale.js:
```diff
@@ -0,0 +1,44 @@
+     * @param  {string | Object} locale The locale in string or object format.
+     * @return {Promise}       Result of formatDefaultLocale call.
+     */
+
```
Nit: No space here, thanks!

maikdiepenbroek replied:
Removed it.

Golodhros commented on src/charts/helpers/locale.js:
```diff
@@ -0,0 +1,44 @@
+     */
+
+    function isValidLocale(locale) {
+        return /[a-z]{2}-[A-Z]{2}/.test(locale);
```
I would move the RegEx as a constant at the top of this file.

maikdiepenbroek replied:
Moved it to a const.

Golodhros commented on src/charts/helpers/locale.js:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    function isValidLocaleDefinition(locale) {
+        const requiredKeys = ['decimal', 'thousands', 'grouping', 'currency']
```
I would move it to a constant at the top of the file as well.

maikdiepenbroek replied:
Moved it to a const

Golodhros commented on src/charts/helpers/locale.js:
```diff
@@ -0,0 +1,44 @@
+        return /[a-z]{2}-[A-Z]{2}/.test(locale);
+    }
+
+    function isValidLocaleDefinition(locale) {
```
Missing a small comment here, thanks!

maikdiepenbroek replied:
Added it

Golodhros commented on test/specs/bar.spec.js:
```diff
@@ -767,6 +767,18 @@ define(['d3', 'bar', 'barChartDataBuilder'], function(d3, chart, dataBuilder) {
+            })
         });
     });
 });
```
I would add another test within the 'render' section, checking that a label is different when using a different locale.

maikdiepenbroek replied:
I'm having trouble getting the test to pass, a little help here would be very much appreciated.

Golodhros replied:
OK, could you upload the test in a commented state?

maikdiepenbroek replied:
Sure, coming up!

maikdiepenbroek replied:
I've added the commented test, been driving me crazy haha. I hope you can take a look. @Golodhros

Golodhros commented on test/specs/helpers.spec.js:
```diff
@@ -441,5 +443,34 @@ define([
+                expectAsync(locale.setDefaultLocale(invalidLocaleDefinition)).toBeRejected();
+                done();
+            });
+        });
```
I would add a couple of test cases for the two OK cases.

maikdiepenbroek replied:
I had those tests, but since i wasn't able to mock the d3.json call, i removed those tests since i would be relying on an internet connection to make the test pass. If that's not a problem, i could re add the OK cases test.

Golodhros replied:
Makes sense, add those here commented for now, I will take a look later.

maikdiepenbroek replied:
Ok, re-added them

Golodhros commented on package.json:
```diff
@@ -80,6 +80,7 @@
     "babel-eslint": "7.2.3",
     "babel-istanbul": "^0.12.2",
     "babel-loader": "7.1.2",
+    "babel-polyfill": "6.26.0",
```
Why did you need this? Only for the tests?

maikdiepenbroek replied:
This was required for the tests indeed, the parsing failed due to missing Promise support.

Golodhros commented on src/charts/bar.js:
```diff
@@ -1292,6 +1300,22 @@ define(function(require) {
+        /**
+         * Gets or Sets the locale which d3-format uses.
+         * (Default is en-US)
+         * @param  {string} _x                   Desired locale
```
You can do [_x = 'es-US'] to mark the default.

Could you add a link in the comment to point to the possible values?
Also, if we accept a locale object instead of the string, we should explain its values or add another link for referencing the format of it.

maikdiepenbroek replied:
Fixed

Golodhros commented on src/charts/bar.js:
```diff
@@ -178,7 +184,9 @@ define(function(require) {
                 if (enableLabels) {
-                    drawLabels();
+                    localeRequest
+                        .then(() => drawLabels())
```
Wouldn't the locale change also the axis tick text?

maikdiepenbroek replied:
I'm not sure to be honest.

Golodhros replied:
No prob, I will test it later.

Golodhros commented on test/specs/bar.spec.js:
```diff
@@ -176,6 +176,44 @@ define(['d3', 'bar', 'barChartDataBuilder'], function(d3, chart, dataBuilder) {
                 });
             });
+
+            // describe('Locale', () => {
```
Nit: you can just add an 'x' as in 'xdescribe' to comment the whole thing.

maikdiepenbroek replied:
True, but we agreed on pushing a commented out test for you to have a look at, right ?

Golodhros replied:
Yeah, I mean that adding the 'x' is how you comment the test. Not a big deal though.

maikdiepenbroek replied:
Fixed

Golodhros commented on src/charts/bar.js:
```diff
@@ -210,28 +218,28 @@ define(function(require) {
                 .append('g')
-                  .classed('container-group', true)
-                  .attr('transform', `translate(${margin.left + yAxisPaddingBetweenChart}, ${margin.top})`);
+                .classed('container-group', true)
```
Oh, what happened here?
append is a special case that gets 2 spaces instead of four.
Was this eslint?

maikdiepenbroek replied:
This happened after i ran `yarn lint:fix`

Golodhros replied:
I see, is there an easy way of rollback that change only? It messed up all the indentation.

maikdiepenbroek replied:
Rolled back before the linting 👍

Golodhros commented:
Awesome work @maikdiepenbroek!!!

Let me play a bit with it and see how we move forward, thanks a lot!

maikdiepenbroek commented:
@Golodhros No problem, been fun and thanks for the patience and collaboration 👍

Golodhros commented:
> @Golodhros No problem, been fun and thanks for the patience and collaboration 👍

On the contrary! Thanks for your patience and sorry for the back and forths!
